### PR TITLE
Fixes #220 - Use article content when a single one exists in the page.

### LINKED ISSUE #220: France24.com missing content
Affected builds:
- Firefox 38.0.5 beta 1

Affected OS`s:
 - Windows 7 64-bit
 - Mac OS X 10.9.5
 - Ubuntu 14.04 32-bit

STR:
1. Visit http://www.france24.com/en/20150512-bangladesh-third-blogger-hacked-death
2. Enter reader view

Expected results:
Content is shown in reader view.

Actual results:
Improper content shown in reader view.

### PR DESCRIPTION
**Pull request by n1k0**
If there's a single `<article>` tag in a page, it's most likely where main contents reside.

r=? @leibovic @gijsk

### CODE REVIEW

File: Readability.js [hunk 1]
```diff
@@ -624,6 +624,11 @@ Readability.prototype = {
 
     var pageCacheHtml = page.innerHTML;
 
+    var articleTags = this._doc.getElementsByTagName("article");
+    if (articleTags.length === 1) {
+      this._doc.body.innerHTML = articleTags[0].parentNode.innerHTML;
```
Comment (gijsk) [inline Readability.js]: There are some issues here.

1. This will break if the page goes through the `while` loop below a second time, as we'll re-set the `innerHTML` to `pageCacheHtml` which is set directly above this block.
2. This leaves a whole lot of work inside the code here, like looking for candidates and whatnot, whereas the summary seems to indicate we just want to use the entire contents of the `article` tag. Why do it this way, then?
3. Why the `parentNode` of the `article`, which will include siblings?
4. This has a performance penalty: we'll be reparsing the HTML. I would suggest:
`this._doc.body.parentNode.replaceChild(articleTags[0].parentNode, this._doc.body)` which is almost O(1).

File: test/test-pages/remove-extra-brs/source.html [hunk 2]
```diff
@@ -11,9 +11,9 @@
       <br>
       <p>Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
         tempor incididunt ut labore et dolore magna aliqua.</p>
-      <br><br><p>Ut enim ad minim veniam,
+      <br><br>Ut enim ad minim veniam,
       quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
-      consequat.</p><br>
+      consequat.<br>
```
Comment (gijsk) [inline test/test-pages/remove-extra-brs/source.html]: Why this change?

Comment (n1k0) [inline test/test-pages/remove-extra-brs/source.html]: Because jsdom was breaking on it.

File: Readability.js (hunk 1 above)
Comment (n1k0) [inline Readability.js]: > 1. This will break if the page goes through the `while` loop below a second time, as we'll re-set the `innerHTML` to `pageCacheHtml` which is set directly above this block.

Good catch.

> 2. This leaves a whole lot of work inside the code here, like looking for candidates and whatnot, whereas the summary seems to indicate we just want to use the entire contents of the `article` tag. Why do it this way, then?

Because it was simple :) I can try skipping the candidate selection part though.

> 3. Why the `parentNode` of the `article`, which will include siblings?

Fair point.

> 4. This has a performance penalty: we'll be reparsing the HTML. I would suggest:
`this._doc.body.parentNode.replaceChild(articleTags[0].parentNode, this._doc.body)` which is almost O(1).

Replacing the `<body>` tag with an `<article>` one? Anyway, that breaks everything.

Comment (gijsk) [inline Readability.js]: >> This has a performance penalty: we'll be reparsing the HTML. I would suggest:
`this._doc.body.parentNode.replaceChild(articleTags[0].parentNode, this._doc.body)` which is almost O(1).

> Replacing the `<body>` tag with an `<article>` one? Anyway, that breaks everything.

Ah, d'oh. It's clear I didn't think this through. In that case, I guess:

`this._doc.body.innerHTML = ''; this._doc.body.appendChild(articleTags[0]);`

That or using `_setNodeTag` to make it a `<body>` again.

Comment (n1k0) [inline Readability.js]: I went with

```js
var articleTags = this._doc.getElementsByTagName("article");
    if (articleTags.length === 1) {
      this._forEachNode(page.children, function(pageChild) {
        page.removeChild(pageChild);
      });
      page.appendChild(articleTags[0]);
    }
```

It seems to work well, though your solution is definitely shorter :)

**Comment by n1k0** [comment]
So I tried skipping the top candidate selection, though that breaks a lot of cases; that part of the code would need some heavy refactoring, but I suggest we play the safest we can here.

File: test/test-pages/wapo-2/expected.html [hunk 3]
```diff
@@ -28,8 +28,4 @@
         <p>“Now it’s hard to see what could persuade the Palestinians” to hold up on their ICC plans, Indyk said. “That has nothing to do with negotiations, but if both sides can’t be persuaded to back down, then they will be on a trajectory that could lead to the collapse of the Palestinian Authority because it can’t pay wages anymore.</p>
         <p>“That could be an issue forced onto the agenda about the same time as a potential nuclear deal.”</p>
     </article>
-    <div class="post-body-sig-line">
-        <a href="http://www.washingtonpost.com/people/steven-mufson"><img class="post-body-headshot-left" src="http://img.washingtonpost.com/wp-apps/imrs.php?src=http://www.washingtonpost.com/blogs/wonkblog/files/2014/07/mufson_steve.jpg&amp;h=180&amp;w=180"></a>
-        <p class="post-body-bio has-photo">Steven Mufson covers the White House. Since joining The Post, he has covered economics, China, foreign policy and energy.</p>
-    </div>
```
Comment (n1k0) [inline test/test-pages/wapo-2/expected.html]: Yeah, we're now missing this, because it's outside the `<article/>` element :/

**Comment by gijsk** [comment]
Closing out old conflicting PRs.
